streams/cluster/src/main/java/com/hortonworks/streamline/streams/cluster/register/impl/AbstractServiceRegisterer.java

```diff
@@ -0,0 +1,124 @@
+/**
+ * Copyright 2017 Hortonworks.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+
+ *   http://www.apache.org/licenses/LICENSE-2.0
+
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+package com.hortonworks.streamline.streams.cluster.register.impl;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.hortonworks.streamline.common.Config;
+import com.hortonworks.streamline.streams.catalog.Cluster;
+import com.hortonworks.streamline.streams.catalog.Component;
+import com.hortonworks.streamline.streams.catalog.Service;
+import com.hortonworks.streamline.streams.catalog.ServiceConfiguration;
+import com.hortonworks.streamline.streams.catalog.configuration.ConfigFileReader;
+import com.hortonworks.streamline.streams.catalog.configuration.ConfigFileType;
+import com.hortonworks.streamline.streams.cluster.discovery.ambari.ConfigFilePattern;
+import com.hortonworks.streamline.streams.cluster.register.ManualServiceRegisterer;
+import com.hortonworks.streamline.streams.cluster.service.EnvironmentService;
+import org.apache.commons.io.FilenameUtils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public abstract class AbstractServiceRegisterer implements ManualServiceRegisterer {
+    protected EnvironmentService environmentService;
+
+    protected abstract String getServiceName();
+
+    protected abstract List<Component> createComponents(Config config, Map<String, String> flattenConfigMap);
+
+    protected abstract boolean validateComponents(List<Component> components);
+
+    protected abstract boolean validateServiceConfigurations(List<ServiceConfiguration> serviceConfigurations);
+
+    protected abstract boolean validateServiceConfiguationsViaFlattenMap(Map<String, String> configMap);
+
+    @Override
+    public void init(EnvironmentService environmentService) {
+        this.environmentService = environmentService;
+    }
+
+    @Override
+    public Service register(Cluster cluster, Config config, List<ConfigFileInfo> configFileInfos) throws IOException {
+        Service service = environmentService.initializeService(cluster, getServiceName());
```

Comment:
It creates new Service entity, which is also bound to specific Cluster. It should be created per each request.